Use both absolute and relative tolerance in testing

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34258 Use both absolute and relative tolerance in testing**

This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.

Differential Revision: [D21110255](https://our.internmc.facebook.com/intern/diff/D21110255)

nairbv committed:
Use both absolute and relative tolerance in testing

[ghstack-poisoned]

nairbv commented:
@ezyang @zou3519 @gchanan There's probably still some cleanup I'll need to do and probably some tests that will fail, so not looking for a thorough review, but would like your thoughts on this general approach.

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
+        torch.bfloat16: 0.016,
+        torch.float: 1.3e-6,
+        torch.double: 5e-7,
+        torch.int: 0
```
Maybe you want to just have all the integral types be zero, and not include them in this map?

nairbv replied:
I actually remapped all integral types to torch.int, but that's probably confusing, so I'll fix that.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
+        if isinstance(x, torch.Tensor):
+            dtype = x.dtype
+        elif isinstance(y, torch.Tensor):
+            dtype = y.dtype
```
This suggests some unpleasant directionality in assertEqual, where assertEqual(x, y) is not the same thing as assertEqual(y, x). It's true that this directionality exists today (where we run y.to(x)), but I'm wondering if we shouldn't be doing something like, setting `compared_dtype` to the lowest precision out of all the dtypes involved.

nairbv replied:
yup, I'm thinking to make this more like _get_default_tolerance from testing/__init__.py (and then getting rid of that one in a future PR).

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
+
+
+    def assertEqual(self, x, y, message='', rtol=None, atol=None, allow_inf=False, exact_dtype=None):
+        # we allow setting an absolute tolerance as a positional arg for BC with legacy testing behavior.
```
It might be easier if you do a preparatory commit before this one that makes everything after `tol` kwarg only, and then fix call sites.

ezyang replied:
This is going to need XLA updates too.

nairbv replied:
actually I did, I forgot that wasn't allowed in python2. I had it working that way, and then changed it back this morning as py2 tests failed.

nairbv replied:
and, yeah I noticed that XLA tests failed. cc @ailzhang

ezyang replied:
If you go kwarg only you're going to have to do it manually; take `**kwargs` and then parse it out. Well, or wait until we kill py2 support.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -777,27 +777,65 @@ def safeCoalesce(self, t):
+    def assertEqual(self, x, y, message='', rtol=None, atol=None, allow_inf=False, exact_dtype=None):
+        # we allow setting an absolute tolerance as a positional arg for BC with legacy testing behavior.
+        if isinstance(message, Number):
+            self.assertIsNone(atol, "don't combine positional prec and atol")
```
If the old user wrote `assertEqual(x, y, tol, "foo bar")` they'll get a weird error message here, right?

Example:
```
test/test_torch.py:2471:        self.assertEqual(target_value, forked_value, 0, "RNG has not forked correctly.")
```

nairbv replied:
That's correct. test_torch.py:2471 was changed to:
        `self.assertEqual(target_value, forked_value, atol=0, message="RNG has not forked correctly.")`
in this PR

Actually though, I'll add a unit test to ensure that calls in that form do in fact fail.

ezyang commented:
I'm very supportive of this change! But I think that it will be easier to land this if you stage it, into making assertEqual arguments kwarg only, and then add your changes.

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv commented:
> I'm very supportive of this change! 

thanks, that's the main thing I'm looking for on the general approach. I have some cleanup to do still I think.

> But I think that it will be easier to land this if you stage it, into making assertEqual arguments kwarg only, and then add your changes.

Do you just mean within this PR, or broken up to multiple PRs? I initially wrote and submitted this as kwarg only and cleaned up issues that broke, but that syntax isn't supported in python 2. When are we actually dropping the python2 tests?

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "[WIP] Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv commented on test/test_torch.py:
```diff
@@ -4323,6 +4323,9 @@ def test_assertEqual(self):
         self.assertEqual(xv, x)
 
+        self.assertRaisesRegex(AssertionError, "don't combine", 
+                               lambda: self.assertEqual(x, xv, 1.0, ""))
```
@ezyang this is the test for your comment about verifying "they'll get a weird error message here, right?"

nairbv commented on torch/testing/_internal/common_utils.py:
```diff
@@ -777,27 +777,67 @@ def safeCoalesce(self, t):
+            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
+                dtype = torch.from_numpy(a).dtype
+
+            # todo: set 0-tolerance for integral type.
```
To do this properly I don't want to have lists of dtypes, and we don't have an `is_integral` in python (`not is_floating_point` will include quantized types). I have a PR here to add is_integral:
https://github.com/pytorch/pytorch/pull/34312
but there are some outstanding questions on the approach, so I'd rather finish this and tighten integral tolerances in a subsequent PR.

ezyang commented:
> Do you just mean within this PR, or broken up to multiple PRs? I initially wrote and submitted this as kwarg only and cleaned up issues that broke, but that syntax isn't supported in python 2. When are we actually dropping the python2 tests?

Yeah, I recommend just taking it as `**kwargs` and then parse things out by hand.

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

anjali411 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -834,30 +888,23 @@ def assertTensorsEqual(a, b):
-                        max_err = diff.max()
-                        self.assertLessEqual(max_err, prec, message)
+                        if a.is_complex():
+                            # todo: assert_allclose should handle complex types directly.
```
as discussed before, we shouldn't compare angles and absolute values, reason being that there might be a substantial difference in the values for same (but very large) absolute value and small angle difference.

I have added the temporary real/imag functions for complex in this PR https://github.com/pytorch/pytorch/pull/35879 so we should instead use that [method calls: ```copy_real(), copy_imag()```]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+        'float32': (1.3e-6, 1e-5),
+        'float64': (1e-7, 1e-7),
+        'complex32': (1.3e-6, 1e-5),
+        'complex64': (1e-7, 1e-7),
```
cc @anjali411 here, it seemed like you needed more tolerance for some complex tests, so it's interesting that the empirical numbers here have complex and float line up exactly.

nairbv replied:
talked to Anjali this should have been 64 and 128 to align with the float types, will fix.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+    # todo: implement numpy-like issubdtype
+    def is_integral(self, dtype):
+        dtypes = get_all_dtypes()
+        return dtype in dtypes and not dtype.is_floating_point
```
The logic here... seems a bit odd? Why would we care if `dtype in dtypes`?

nairbv replied:
this is just a hack until we have an issubdtype api, get_all_dtypes doens't have certain dtypes like quantized or complex that aren't integral.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+        dtypes = get_all_dtypes()
+        return dtype in dtypes and not dtype.is_floating_point
+
+    def get_default_tolerance(self, a, b=None):
```
What are `a` and `b` here?

nairbv replied:
either tensors or dtypes or ndarray. I'll add a comment.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+            elif isinstance(a, torch.dtype):
+                dtype = a
+            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
+                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
```
What's up with the special case here?

nairbv replied:
'U' is unicode. I'll add a comment. I forget what test broke without this, but for some reason we do have cases where we've passed unicode numpy arrays to assertEquals

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+            if self.is_integral(dtype):
+                return (0, 0)
+            dtype = str(dtype).split('.')[-1]
+            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
```
This is not a big deal, but personally I would have just keyed `dtype_precisions` directly on the dtype objects themselves. Then you don't have to stringify and munge it

nairbv replied:
yeah I initially did too, then copied an approach used elsewhere in our code. I had been thinking I might want to include other non-dtype names like 'integral'. I might make some small adjustments in a followup PR once the bulk of this is landed.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -789,27 +790,80 @@ def safeCoalesce(self, t):
+        rtol = kwargs.get('rtol', None)
+        atol = kwargs.get('atol', None)
+        allow_inf = kwargs.get('allow_inf', False)
+        exact_dtype = kwargs.get('exact_dtype', None)
```
It would be good to do some sort of test to make sure someone doesn't send a `kwarg` which we ignore. But there's even a better solution now that we're Python 3 only: use keyword only arguments `(self, x, y, *, rtol=..., ....)`

nairbv replied:
ah when I wrote this we still had python2 support. I think I might also prefer to land that update separately just to minimize conflicts in the meantime.

ezyang commented on torch/testing/_internal/common_utils.py:
```diff
@@ -956,14 +996,14 @@ def assertNotEqual(self, x, y, prec=None, message=''):
                 # https://github.com/pytorch/pytorch/issues/22301
                 max_err = diff.max().item()
-                self.assertGreaterEqual(max_err, prec, message)
+                self.assertGreater(max_err, atol, message)
```
Any particular reason for the change here? Seems subtle.

nairbv replied:
assertGreaterEqual won't work when using a tolerance of zero (exact match).
e.g. in `assertNotEqual( 5, 5, atol=0 )`, if you're doing `assertGreaterEqual(5-5, 0)` you're going to pass when you should fail.

ezyang reviewed (approved):
Since this PR involves a decently large codemod, let's err on the side of merging first and asking questions later.

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

nairbv committed:
Update on "Use both absolute and relative tolerance in testing"


This PR allows both atol and rtol to be specified, uses defaults based on the prior analysis (spreadsheet attached to https://github.com/pytorch/pytorch/pull/32538), but retains the absolute tolerance behavior in cases where precision was previously specified explicitly.


[ghstack-poisoned]

ailzhang commented:
@nairbv This PR didn't fix all places that uses `assertEqual` like we pass float/int as `message`. E.g. https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L11037 
and https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L11001 by a quick look. 
@nairbv Would you mind giving it another pass to fix (potentially more) uses? Otherwise we'll get weird error messages when tests fail.

nairbv commented:
> @nairbv This PR didn't fix all places that uses `assertEqual` like we pass float/int as `message`. E.g. https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L11037
> and https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L11001 by a quick look.
> @nairbv Would you mind giving it another pass to fix (potentially more) uses? Otherwise we'll get weird error messages when tests fail.

Hmm... Shouldn't this code in assertEqual handle it? I was thinking maybe you found some int vs float issue or something, but those are both instances of `Number`. I'm not sure I understand the case you're seeing.

        if isinstance(message, Number):
            self.assertIsNone(atol, "don't combine positional prec and atol")
            self.assertIsNone(rtol, "don't combine positionial prec and rtol")
            atol = message
            message = ''
            rtol = 0

ailzhang commented:
@nairbv Aha I see, but IMHO `if isinstance(message, Number):` is not a common logic. 
- Like if I were a new dev, I wouldn't expect passing an int/float to `message` to work until I see this implementation, but it magically works. 
- And for old devs, it's confusing whether the third positional arg should be a message(str) or a precision(number) (we support both, but it's too magical especially when we call that arg `message`). 
- Also for downstream `pytorch/xla` where we want to override `atol` etc, now we have two potential sources to override. One is kwarg `atol`, the other is `message`. 

Hmmm I personally feel fixing callsites and make kwarg `message` only handle `str` is more straightforward.

nairbv commented:
> Hmmm I personally feel fixing callsites and make kwarg `message` only handle `str` is more straightforward.

Yup I think not-allowing numbers is preferable, and I'd generally discourage the use of this field. For a first pass I was just trying to keep it simple and semi-BC to not have to change quite as many tests.

We can add an assertion of `not isinstance(message, Number)` and fix breakages. I don't think anything is technically incorrect for now though right?

ailzhang commented:
@nairbv Yup that'll work! Yea totally makes sense that it can be a followup. It's not wrong but just a little confusing. Thanks!!

xwang233 commented:
Hi, I saw that `allow_inf` argument is no longer used when both x and y are tensor. Should that be (fixed and) passed here, or add some additional logic?
https://github.com/pytorch/pytorch/blob/3d898fa76360fa6e4f572caa5ab54792d8b9ea28/torch/testing/_internal/common_utils.py#L909-L910

nairbv commented:
> Hi, I saw that `allow_inf` argument is no longer used in `self.assertEqual` when both x and y are tensor. Should that be (fixed and) passed here, or add some additional logic?
> https://github.com/pytorch/pytorch/blob/3d898fa76360fa6e4f572caa5ab54792d8b9ea28/torch/testing/_internal/common_utils.py#L909-L910

ah, I think you're correct, we're handling the False case for Numbers but not Tensors now. For allow_inf=False we should be doing some kind of assertion that the tensor does not contain inf values. Probably something like:
```
if not allow_inf:
    assertTrue(torch.all(~ torch.isinf(a)))
```

The existing mask code was incorrect (e.g. it's trying to take the sign of a boolean), but assert_allclose allows inf by default and validates that the sign of the inf's match.
